datafusion/optimizer/src/simplify_expressions/expr_simplifier.rs

```diff
@@ -365,6 +367,32 @@ impl<'a, S: SimplifyInfo> ExprRewriter for Simplifier<'a, S> {
                     None => lit_bool_null(),
                 }
             }
+            Expr::InList {
+                expr: _expr,
+                list,
+                negated,
+            } if list.is_empty() => lit(negated),
+
+            Expr::InList {
+                expr,
+                list,
+                negated,
+            } if list.len() <= THRESHOLD_INLINE_INLIST && expr.try_into_col().is_ok() => {
+                let first_val = list[0].clone();
+                if negated {
+                    list.into_iter()
+                        .skip(1)
+                        .fold((*expr.clone()).not_eq(first_val), |acc, y| {
+                            (*expr.clone()).not_eq(y.clone()).and(acc)
+                        })
+                } else {
+                    list.into_iter()
+                        .skip(1)
+                        .fold((*expr.clone()).eq(first_val), |acc, y| {
+                            (*expr.clone()).eq(y.clone()).or(acc)
+                        })
+                }
```

Comment:
I think you can simplify this code using `reduce`:

```suggestion
                if negated {
                    list.into_iter()
                        .reduce(|acc, y| {
                            (*expr.clone()).not_eq(y.clone()).and(acc)
                        }).unwrap()
                } else {
                    list.into_iter()
                        .reduce(|acc, y| {
                            (*expr.clone()).eq(y.clone()).or(acc)
                        }).unwrap()
                }
```

Reply:
Unfortunately I don't think it's possible (this was my first try as well), as we also need to transform the first `acc` value.
Here the first acc won't be of `expr.eq(acc)` but just `acc`.